### PR TITLE
add option to promtool to write JUnit-style XML test results

### PR DESCRIPTION
**Pull request by maokt**
Signed-off-by: Marty Pauley <marty@martian.org>

### CODE REVIEW

**Comment by maokt** [comment]
@dgl, @roidelapluie, I would like to add a "--junit=" option to promtool's "test" subcommand to write an XML file with the test results in "JUnit" format, because we need to record evidence of testing in some standard machine-parseable format.

File: cmd/promtool/unittest.go [hunk 1]
```diff
@@ -35,15 +36,24 @@ import (
 	"github.com/prometheus/prometheus/promql/parser"
 	"github.com/prometheus/prometheus/rules"
 	"github.com/prometheus/prometheus/storage"
+
+	"github.com/maokt/junitxml"
```
Comment (dgl) [inline cmd/promtool/unittest.go:40]: Given how simple this is I'd prefer to just add this as another file here for now, makes maintenance easier.

Comment (roidelapluie) [reply cmd/promtool/unittest.go:40]: this could also be a package in utils/

Comment (maokt) [reply cmd/promtool/unittest.go]: ok, will do that

**Comment by maokt** [comment]
Done, I think. Will sleep while it's building...
Thanks.

**Comment by maokt** [comment]
I'm assuming there's nothing I can do about those 5 failing checks as they're all about react.

**Comment by maokt** [comment]
> I'm assuming there's nothing I can do about those 5 failing checks as they're all about react.

Hmm... maybe I can. Will rebase this and try again...

**Comment by maokt** [comment]
resolves #12253

**Comment by maokt** [comment]
@dgl I would be grateful if you could review this again. I think it would resolve at least two open issues requesting the feature. Thanks.

**Comment by dgl** [comment]
Sorry, on my list, just a bit busy atm!

**Comment by maokt** [comment]
Thanks.

File: util/junitxml/junitxml_test.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
+// Copyright 2021 Marty Pauley
```
Comment (roidelapluie) [inline util/junitxml/junitxml_test.go:1]: We set the header to "The prometheus authors" so we don't have to change on every edit

File: util/junitxml/junitxml.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
+// Copyright 2021 Marty Pauley
```
Comment (roidelapluie) [inline util/junitxml/junitxml.go:1]: We set the header to "The prometheus authors" so we don't have to change on every edit

**Comment by bboreham** [comment]
Discussed at the bug-scrub.
@maokt for most 1-line changes to comments I would just do it, but removing your name from a Copyright message seems more sensitive.  If you can make the change we can merge.

**Comment by beorn7** [comment]
Hello from the bug scrub.

With no response from the original author, we have to close this. See #8860 for next steps.

**Comment by michal-sankot** [comment]
Why exactly is it closed? We would definitely use this feature as we run tests during deployment in Jenkins.

**Comment by beorn7** [comment]
As said above, that's explained [here](https://github.com/prometheus/prometheus/issues/8860#issuecomment-2175918745).

**Comment by michal-sankot** [comment]
Ok, I see. So does it need an author to fix it with a new PR or anybody can do it? (If it's just a formal thing I can do it to make it move ahead).

**Comment by beorn7** [comment]
We just need somebody who implements the same thing as in this PR (but in their own words). It should be easy, a very nice "first issue in Prometheus". If you want to do it, you are very welcome. (Best to say so on the issue so we avoid redundant work.)

**Comment by beorn7** [comment]
@michal-sankot I just got message that somebody is already working on the PR.

**Pull request closed without merging**